### DISCUSSION TITLE
Conflicting keymaps

### DISCUSSION THREAD

**Discussion by danpascu** [LazyVim/LazyVim#3863] [Q&A]
I noticed there are a lot of conflicting keymaps.

which-key reports these during checkhealth:

```
which-key: require("which-key.health").check()

WhichKey: checking conflicting keymaps ~
- WARNING conflicting keymap exists for mode **"o"**, lhs: **"a"**
- rhs: ` `
- WARNING conflicting keymap exists for mode **"o"**, lhs: **"an"**
- rhs: ` `
- WARNING conflicting keymap exists for mode **"o"**, lhs: **"al"**
- rhs: ` `
- WARNING conflicting keymap exists for mode **"o"**, lhs: **"i"**
- rhs: ` `
- WARNING conflicting keymap exists for mode **"o"**, lhs: **"in"**
- rhs: ` `
- WARNING conflicting keymap exists for mode **"o"**, lhs: **"il"**
- rhs: ` `
- WARNING conflicting keymap exists for mode **"n"**, lhs: **"gc"**
- rhs: ` `
- WARNING conflicting keymap exists for mode **"n"**, lhs: **"y"**
- rhs: `<Plug>(YankyYank)`
- WARNING conflicting keymap exists for mode **"v"**, lhs: **"a"**
- rhs: ` `
- WARNING conflicting keymap exists for mode **"v"**, lhs: **"al"**
- rhs: ` `
- WARNING conflicting keymap exists for mode **"v"**, lhs: **"an"**
- rhs: ` `
- WARNING conflicting keymap exists for mode **"v"**, lhs: **"i"**
- rhs: ` `
- WARNING conflicting keymap exists for mode **"v"**, lhs: **"il"**
- rhs: ` `
- WARNING conflicting keymap exists for mode **"v"**, lhs: **"in"**
- rhs: ` `
```

None of these are added in my configuration.

But I noticed there are even more that are not reported. For example if I quickly press `gr` without any pause between the letters, I get a Telescope view of the lsp references for a symbol. But if I press `g` and wait until the which-key popup appears to show the available continuations and then press `r`, it leads to a submenu that has 3 entries:

```
a -> vim.lsp.buf.code_action()
n -> vim.lsp.buf.rename()
r -> vim.lsp.buf.references()
```

I ran `:nmap` and looked through the keymaps to find these entries related to `gr`:

```
n  gr          *@<Cmd>Telescope lsp_references<CR>

n  grÞ         * <Nop>
n  gr          * <Cmd>lua require("which-key").show("gr", {mode = "n", auto = true})<CR>
n  gÞ          * <Nop>
n  g           * <Cmd>lua require("which-key").show("g", {mode = "n", auto = true})<CR>

n  grr         * <Lua 17: vim/_defaults.lua:0>
                 vim.lsp.buf.references()
n  gra         * <Lua 15: vim/_defaults.lua:0>
                 vim.lsp.buf.code_action()
n  grn         * <Lua 14: vim/_defaults.lua:0>
                 vim.lsp.buf.rename()
```

I believe the `gra`, `grn` and `grr` keymaps are predefined by nvim itself for its lsp client, while `gr` comes from the telescope plugin.

This is with nvim version NVIM v0.11.0-dev-138+g88fe467b1, not sure if that makes a difference.

_Originally posted by @danpascu in https://github.com/LazyVim/LazyVim/discussions/3812_

**Comment by folke** [comment]
It's a **warning**. And next time, please oh please use discussions for user related questions.

**Reply by danpascu** [reply]
I used a discussion, and there was no reply. I converted to an issue because I think it is a bug to have overlapping keymaps.

Also, you claim that the fact that depending on how fast or slow I press `gr` I end in 2 different places is a warning and not a bug? I'm confused.

I'm not really asking a question here, I'm reporting something that's problematic  (`gr` -> telescope, `g <pause> r` -> lsp sub-menu).

**Reply by dpetka2001** [reply]
These mappings came to life with Neovim 0.10. There were not there before and most distros and other LSP plugins have been using these keymaps as is right now. Please read `:h lsp-defaults`. Decide which ones you want and delete the ones that you don't want. Using `vim.keymap.del` you can delete the ones you don't want. And having overlapping keymaps is definitely not a bug. It's just an inconvenience and users are free to change the keymaps to the ones they want to use. Not everyone's preferences can be taken into account neither on a Neovim distro nor Neovim defaults.

**Reply by stonemaster** [reply]
Actually coming back to this, `gr` is used by the Octo plugin when installed using `util.octo`: [1]

```lua
[...]
    { "<leader>gr", "<cmd>Octo repo list<CR>", desc = "List Repos (Octo)" },
```

I am not sure whether this is a real bug, but from a user perspective having Octo installed leads to a confusing behaviour because it overwrites the default LSP references keymap.

[1]: https://www.lazyvim.org/extras/util/octo

**Reply by dpetka2001** [reply]
No, it does not overwrite anything. The keymap Octo is using is `<leader>gr` which is different from plain `gr`, so no problem.

**Reply by stonemaster** [reply]
Ah you are right, my mistake. Thanks for the clarification!
